### PR TITLE
Improve `optimizeDeps.entries` to avoid server endpoints

### PR DESCRIPTION
**Pull request by bluwy**
## Changes

close https://github.com/withastro/astro/issues/9859

Previously, `optimizeDeps.entries` set does not respect the `srcDir` option and globs into server endpoints, which we know should not contain browser code. `optimizeDeps` should only be used on browser code.

This PR fixes by respecting `srcDir`, and make sure server endpoints are not included in the glob. While this is not perfect in the bigger picture, and the ideal solution is to add a custom esbuild plugin to transform Astro files to allow the scanner to scan the appropriate browser files, this PR is an easier and quicker fix to the issue for now.

## Testing

It's a bit hard to test the overall behaviour, but I ran `DEBUG="vite:deps" pnpm dev` in the example projects to check that it logs the correct files globbed by `optimizeDeps.entries`.

## Docs

n/a. bug fix.

### CODE REVIEW

File: packages/astro/src/core/create-vite.ts [hunk 1]
```diff
@@ -112,7 +115,13 @@ export async function createVite(
 		customLogger: createViteLogger(logger, settings.config.vite.logLevel),
 		appType: 'custom',
 		optimizeDeps: {
-			entries: ['src/**/*'],
+			// Scan all files within `srcDir` except for known server-code (e.g endpoints)
+			entries: [
+				`${srcDirPattern}!(pages)/**/*`, // All files except for pages
```
Comment (bluwy) [inline packages/astro/src/core/create-vite.ts:120]: I made some tests and it seems that `srcDirPattern` will always have a trailing slash. The only case I'm not sure is on Windows.

File: packages/astro/src/core/create-vite.ts [hunk 2]
```diff
@@ -112,7 +115,13 @@ export async function createVite(
...
+				`${srcDirPattern}pages/**/!(*.js|*.ts)`, // All pages except for endpoints
```
Comment (ematipico) [inline packages/astro/src/core/create-vite.ts:121]: Ehehe, unfortunately `mjs` and `cjs` should be included, even though it's a low chance that users use them. Still, we should

**Review by ematipico — APPROVED** [review]

File: packages/astro/src/core/create-vite.ts (hunk 2 above)
Comment (bluwy) [reply packages/astro/src/core/create-vite.ts:121]: Good catch! I'll add this too

**Review by ematipico — APPROVED** [review]
Looks good!

**Pull request merged**